File: algo/56. Merge Intervals.py
# Definition for an interval.
# class Interval:
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution:
    def merge(self, intervals: 'List[Interval]') -> 'List[Interval]':
        helper_dict = {}
        for interval in intervals:
            if interval.start not in helper_dict.keys():
                helper_dict[interval.start] = interval
            else:
                if interval.end > helper_dict[interval.start].end:
                    helper_dict[interval.start] = interval
                else:
                    continue
        helper_dict = sorted(helper_dict.items())
        res = []
        for i in range(len(helper_dict)):
            currInterval = helper_dict[i][1]
            if i == 0:
                res.append(currInterval)
            else:
                lastInterval = res[-1]
                if currInterval.start> lastInterval.end:
                    res.append(currInterval)
                else:
                    res[-1].end = max(lastInterval.end, currInterval.end)
        return res

File: round3/56. Merge Intervals.cs
public class Solution {
    public int[][] Merge(int[][] intervals) {
        Array.Sort(intervals, (a, b) => {return a[0] - b[0];});
        
        var lst = new List<int[]>();
                
        int[] curr = null;
        
        for (int i = 0; i < intervals.Length; i ++) {
            if (curr == null) {
                curr = intervals[i];
            } else {
                if (curr[1] < intervals[i][0]) {
                    lst.Add(curr);
                    curr = intervals[i];
                } else {
                    curr[1] = Math.Max(curr[1], intervals[i][1]);
                }
            }
        }
        
        if (curr != null) lst.Add(curr);;
        
        return lst.ToArray();
    }
}